src/web-push-client.js

```diff
@@ -0,0 +1,329 @@
+      );
+    }
+
+    if (!window.isSecureContext) {
+      throw new Error(
+        'Pusher Beams relies on Service Workers, which only work in secure contexts. Check that your page is being served from localhost/over HTTPS'
+      );
+    }
+
+    if (!('serviceWorker' in navigator)) {
+      throw new Error(
+        'Pusher Beams does not support this browser version (Service Workers not supported)'
+      );
+    }
+
+    if (!('PushManager' in window)) {
+      throw new Error(
+        'Pusher Beams does not support this browser version (Web Push not supported)'
+      );
+    }
+
+    if (serviceWorkerRegistration) {
+      const serviceWorkerScope = serviceWorkerRegistration.scope;
+      const currentURL = window.location.href;
+      const scopeMatchesCurrentPage = currentURL.startsWith(serviceWorkerScope);
+      if (!scopeMatchesCurrentPage) {
+        throw new Error(
+          `Could not initialize Pusher web push: current page not in serviceWorkerRegistration scope (${serviceWorkerScope})`
+        );
+      }
+    }
+
+    this.instanceId = instanceId;
+    this._deviceId = null;
+    this._token = null;
+    this._userId = null;
+    this._serviceWorkerRegistration = serviceWorkerRegistration;
+    this._deviceStateStore = new DeviceStateStore(instanceId);
+    this._endpoint = endpointOverride; // Internal only
+    this._platform = 'web';
+
+    this._ready = this._init();
+  }
+
+  async _init() {
+    if (this._deviceId !== null) {
+      return;
+    }
+
+    await this._deviceStateStore.connect();
+
+    if (this._serviceWorkerRegistration) {
+      // If we have been given a service worker, wait for it to be ready
+      await window.navigator.serviceWorker.ready;
+    } else {
+      // Otherwise register our own one
+      this._serviceWorkerRegistration = await getServiceWorkerRegistration();
+    }
+
+    await this._detectSubscriptionChange();
+
+    this._deviceId = await this._deviceStateStore.getDeviceId();
+    this._token = await this._deviceStateStore.getToken();
+    this._userId = await this._deviceStateStore.getUserId();
+  }
+
+  async _detectSubscriptionChange() {
+    const storedToken = await this._deviceStateStore.getToken();
+    const actualToken = await getWebPushToken(this._serviceWorkerRegistration);
+
+    const pushTokenHasChanged = storedToken !== actualToken;
+
+    if (pushTokenHasChanged) {
+      // The web push subscription has changed out from underneath us.
+      // This can happen when the user disables the web push permission
+      // (potentially also renabling it, thereby changing the token)
+      //
+      // This means the SDK has effectively been stopped, so we should update
+      // the SDK state to reflect that.
+      await this._deviceStateStore.clear();
+      this._deviceId = null;
+      this._token = null;
+      this._userId = null;
+    }
+  }
+
+  async start() {
+    await this._resolveSDKState();
```

Comment:
Maybe we should remove this call here, which detects subscription changes, so that the behaviour is consistent with Safari. Otherwise, developers might get bugs where the state SDK is different when running in Safari vs other browsers.